src/ui/public/persisted_state/persisted_state.js

```diff
@@ -0,0 +1,194 @@
+    };
+
+    PersistedState.prototype.fromString = function (input) {
+      return this.set(JSON.parse(input));
+    };
+
+    PersistedState.prototype._getIndex = function (key) {
+      if (_.isUndefined(key)) return this._path;
+      return (this._path || []).concat(toPath(key));
+    };
+
+    PersistedState.prototype._getDefault = function () {
+      var def = (this._hasPath()) ? undefined : {};
+      return (this._parent ? this.get() : def);
+    };
+
+    PersistedState.prototype._setPath = function (path) {
+      var isString = _.isString(path);
+      var isArray = _.isArray(path);
+
+      if (!isString && !isArray) return [];
+      return (isString) ? [this._getIndex(path)] : path;
+    };
+
+    PersistedState.prototype._hasPath = function () {
+      return this._path.length > 0;
+    };
+
+    PersistedState.prototype._get = function (key, def) {
+      // delegate to parent instance
+      if (this._parent) return this._parent._get(this._getIndex(key), key);
+
+      // no path and no key, get the whole state
+      if (!this._hasPath() && _.isUndefined(key)) {
+        return this._mergedState;
+      }
+
+      return _.get(this._mergedState, this._getIndex(key), def);
+    };
+
+    PersistedState.prototype._set = function (key, value, initialChildState, defaultChildState) {
+      var self = this;
+      var initialState = !this._initialized;
+      var keyPath = this._getIndex(key);
+
+      // if this is the initial state value, save value as the default
+      if (initialState) {
+        this._changedState = {};
+        if (!this._hasPath() && _.isUndefined(key)) this._defaultState = value;
+        else this._defaultState = _.set({}, keyPath, value);
+      }
+
+      // delegate to parent instance, passing child's default value
+      if (this._parent) {
+        return this._parent._set(keyPath, value, initialState, this._defaultState);
+      }
+
+      // everything in here affects only the parent state
+      if (!initialState) {
+        // no path and no key, set the whole state
+        if (!this._hasPath() && _.isUndefined(key)) {
+          // check for changes and emit an event when found
+          if (!_.isEqual(this._changedState, value)) this.emit('change');
+          if (!initialChildState) this._changedState = value;
+        } else {
+          // check for changes and emit an event when found
+          if (!_.isEqual(this.get(keyPath), value)) this.emit('change');
+
+          // arrays merge by index, not the desired behavior - ensure they are replaced
+          if (!initialChildState) {
+            if (_.isArray(_.get(this._mergedState, keyPath))) {
+              _.set(this._mergedState, keyPath, undefined);
+            }
+            _.set(this._changedState, keyPath, value);
+          }
+        }
+
+      }
+
+      var targetObj = this._mergedState || _.cloneDeep(this._defaultState);
+      var sourceObj = _.merge({}, defaultChildState, this._changedState);
```

Comment:
Why does the `_set` method, which should only be executed on the parent, need to know about the `defaultChildState`?